src/resource.rs

```diff
@@ -88,11 +89,27 @@ pub enum WriteError {
 pub trait WriteAtomic: io::Write {
     /// Atomic non-blocking I/O write operation, which must either write the whole buffer to a
     /// resource without blocking - or fail with [`WriteError::NotReady`] error.
+    ///
+    /// # Safety
+    ///
+    /// Panics on invalid [`WriteAtomic::write_or_buf`] implementation, i.e. if it doesn't handle
+    /// EGAGAIN, EINTER, EWOULDBLOCK I/O errors by buffering the data and returns them instead.
     fn write_atomic(&mut self, buf: &[u8]) -> Result<(), WriteError> {
         if !self.is_ready_to_write() {
             Err(WriteError::NotReady)
         } else {
-            self.write_or_buf(buf).map_err(WriteError::from)
+            // TODO: on EGAGAIN, EINTER, EWOULDBLOCK just keep the data buffered
```

Comment:
Is it a problem that we don't do this atm?

By the way `EINTR` can be retried immediately (in a loop). That's how `io::copy` works for eg.

Reply:
This is not a write operation, but a buffered write, meaning that it is the job of the resource to buffer on such errors and not to propagate them (it is the current behavior we have in `netservices`